optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,153 @@
+from optuna import logging
+from optuna.pruners.base import BasePruner
+from optuna.pruners.successive_halving import SuccessiveHalvingPruner
+from optuna import type_checking
+
+if type_checking.TYPE_CHECKING:
+    from typing import List  # NOQA
+
+    from optuna import structs  # NOQA
+    from optuna.trial import Trial  # NOQA
+    from optuna.study import Study  # NOQA
+
+_logger = logging.get_logger(__name__)
+
+
+class HyperbandPruner(BasePruner):
+    """Pruner using Hyperband.
+
+    As SuccessiveHalving (SHA) requires the number of configurations
+    :math:`n` as its hyperparameter.  For a given finite budget :math:`B`,
+    all the configurations have the resources of :math:`B \\over n` on average.
+    As you can see, there will be a trade-off of :math:`B` and :math:`B \\over n`.
+    `Hyperband <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_ attacks this trade-off
+    by trying different :math:`n` values for a fixed budget.
+    Note that this implementation does not take as inputs the maximum amount of resource to
+    a single SHA noted as :math:`R` in the paper.
+
+    Args:
+        min_resource:
+            A parameter for specifying the minimum resource allocated to a trial noted as :math:`r`
+            in the paper.
+            See the details for :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        reduction_factor:
+            A parameter for specifying reduction factor of promotable trials noted as
+            :math:`\\eta` in the paper. See the details for
+            :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        min_early_stopping_rate_low:
+            The start point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+        min_early_stopping_rate_high:
+            The end point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
```

Comment:
How about `eta` following the paper's notation?
```suggestion
            eta=3,
```

Reply:
[offline discussion] for the compatibility with `SuccessiveHalvingPruner`, this should remain as is.